homeassistant/components/nobo_hub/config_flow.py

```diff
@@ -0,0 +1,222 @@
+"""Config flow for Nobø Ecohub integration."""
+from __future__ import annotations
+
+import logging
+import socket
+from typing import Any
+
+from pynobo import nobo
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.const import CONF_COMMAND_OFF, CONF_COMMAND_ON, CONF_IP_ADDRESS
+from homeassistant.core import callback
+from homeassistant.data_entry_flow import FlowResult
+from homeassistant.exceptions import HomeAssistantError
+
+from .const import (
+    CONF_OVERRIDE_TYPE,
+    CONF_OVERRIDE_TYPE_CONSTANT,
+    CONF_OVERRIDE_TYPE_NOW,
+    CONF_SERIAL,
+    CONF_WEEK_PROFILE_NONE,
+    DOMAIN,
+)
+
+DATA_NOBO_HUB_IMPL = "nobo_hub_flow_implementation"
+DEVICE_INPUT = "device_input"
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Nobø Ecohub."""
+
+    VERSION = 1
+
+    def __init__(self):
+        """Initialize the config flow."""
+        self.discovered_hubs = None
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle the initial step."""
+        if self.discovered_hubs is None:
+            self.discovered_hubs = await nobo.async_discover_hubs(loop=self.hass.loop)
+
+        errors = {}
+        if user_input is not None:
+            try:
+                serial = user_input.get(CONF_SERIAL)
+                ip_address = user_input.get(CONF_IP_ADDRESS)
+                if serial is None or not len(serial) == 12 or not serial.isdigit():
+                    raise InvalidSerial()
+                if ip_address is not None:
+                    try:
+                        socket.inet_aton(ip_address)
+                    except OSError:
+                        raise InvalidIP() from OSError
+                else:
+                    for (discovered_ip, serial_prefix) in self.discovered_hubs:
+                        if serial.startswith(serial_prefix):
+                            ip_address = discovered_ip
+                            break
+                    if ip_address is None:
+                        raise DeviceNotFound()
+
+                await self.async_set_unique_id(serial)
+                self._abort_if_unique_id_configured()
+
+                # Test connection
+                hub = nobo(
+                    serial=serial, ip=ip_address, discover=False, loop=self.hass.loop
+                )
+                if await hub.async_connect_hub(ip_address, serial):
+                    await hub.close()
+                    await self.async_set_unique_id(serial, raise_on_progress=False)
+                    self._abort_if_unique_id_configured(
+                        reload_on_update=False, updates=user_input
+                    )
+                    return self.async_create_entry(
+                        title=hub.hub_info["name"], data=user_input
+                    )
+                raise CannotConnect()
```

Comment:
Consider moving this to a validate_input function

Reply:
Good idea. I named it `_test_connection`, as it actually tries to connect, not only validates the user input.